File: sdk/python/term_sdk/runner.py
"""
Agent runner for Term Challenge with HTTP server for persistence.
"""

import sys
import json
import traceback
import time
import signal
from http.server import HTTPServer, BaseHTTPRequestHandler
from .types import Request, Response
from .agent import Agent


# Default port for HTTP server
DEFAULT_PORT = 8765

# Global agent reference for HTTP handler
_agent: Agent = None
_log_enabled = True


def set_logging(enabled: bool) -> None:
    """Enable or disable agent logging."""
    global _log_enabled
    _log_enabled = enabled


def log(msg: str) -> None:
    """Log to stderr (stdout is reserved for protocol)."""
    if _log_enabled:
        print(f"[agent] {msg}", file=sys.stderr, flush=True)


def log_error(msg: str) -> None:
    """Log an error message."""
    print(f"[agent] ERROR: {msg}", file=sys.stderr, flush=True)


class AgentHandler(BaseHTTPRequestHandler):
    """HTTP request handler for agent communication."""
    
    def log_message(self, format, *args):
        """Suppress default HTTP logging."""
        pass
    
    def do_POST(self):
        """Handle POST /step requests."""
        global _agent
        
        if self.path == '/step':
            try:
                content_length = int(self.headers.get('Content-Length', 0))
                body = self.rfile.read(content_length).decode('utf-8')
                
                # Parse request
                request = Request.parse(body)
                instruction_preview = request.instruction[:50].replace('\n', ' ')
                log(f"Step {request.step}: {instruction_preview}...")
                
                # Solve
                start_time = time.time()
                response = _agent.solve(request)
                elapsed_ms = int((time.time() - start_time) * 1000)
                
                # Log response
                if response.command:
                    cmd_preview = response.command[:60]
                    log(f"  -> {cmd_preview}{'...' if len(response.command) > 60 else ''} ({elapsed_ms}ms)")
                if response.task_complete:
                    log(f"  -> Task complete")
                
                # Send response
                response_json = response.to_json()
                self.send_response(200)
                self.send_header('Content-Type', 'application/json')
                self.send_header('Content-Length', len(response_json))
                self.end_headers()
                self.wfile.write(response_json.encode('utf-8'))
                
            except Exception as e:
                log_error(f"Error: {e}")
                traceback.print_exc(file=sys.stderr)
                error_response = Response.done().to_json()
                self.send_response(500)
                self.send_header('Content-Type', 'application/json')
                self.send_header('Content-Length', len(error_response))
                self.end_headers()
                self.wfile.write(error_response.encode('utf-8'))
        
        elif self.path == '/health':
            self.send_response(200)
            self.send_header('Content-Type', 'text/plain')
            self.end_headers()
            self.wfile.write(b'ok')
        
        elif self.path == '/shutdown':
            log("Shutdown requested")
            self.send_response(200)
            self.end_headers()
            # Signal server to stop
            raise KeyboardInterrupt()
        
        else:
            self.send_response(404)
            self.end_headers()
    
    def do_GET(self):
        """Handle GET /health for readiness check."""
        if self.path == '/health':
            self.send_response(200)
            self.send_header('Content-Type', 'text/plain')
            self.end_headers()
            self.wfile.write(b'ok')
        else:
            self.send_response(404)
            self.end_headers()


def run(agent: Agent, port: int = None) -> None:
    """
    Run an agent as HTTP server for the Term Challenge harness.
    
    The agent starts once, handles multiple step requests via HTTP POST /step,
    and maintains state across all steps in a task.
    
    Args:
        agent: Your agent instance
        port: HTTP port (default: 8765, or AGENT_PORT env var)
    
    Example:
        ```python
        from term_sdk import Agent, Request, Response, run
        
        class MyAgent(Agent):
            def setup(self):
                self.memory = []  # Persists across steps!
            
            def solve(self, req: Request) -> Response:
                self.memory.append(req.step)
                return Response.cmd("ls")
        
        if __name__ == "__main__":
            run(MyAgent())
        ```
    """
    global _agent
    
    import os
    if port is None:
        port = int(os.environ.get('AGENT_PORT', DEFAULT_PORT))
    
    _agent = agent
    
    try:
        # Setup agent ONCE at startup
        log("Initializing agent...")
        agent.setup()
        log("Agent ready")
        
        # Start HTTP server
        server = HTTPServer(('0.0.0.0', port), AgentHandler)
        log(f"Listening on port {port}")
        
        # Handle graceful shutdown
        def shutdown_handler(signum, frame):
            log("Received shutdown signal")
            server.shutdown()
        
        signal.signal(signal.SIGTERM, shutdown_handler)
        signal.signal(signal.SIGINT, shutdown_handler)
        
        # Serve forever (until shutdown)
        server.serve_forever()
        
    except KeyboardInterrupt:
        log("Shutting down...")
    except Exception as e:
        log_error(f"Fatal error: {e}")
        traceback.print_exc(file=sys.stderr)
    finally:
        # Cleanup ONCE at end
        log("Cleaning up...")
        agent.cleanup()
        log("Agent finished")


# Legacy stdin/stdout mode for compatibility
def run_stdio(agent: Agent) -> None:
    """
    Run agent in stdin/stdout mode (legacy, single request).
    """
    try:
        agent.setup()
        
        for line in sys.stdin:
            line = line.strip()
            if not line:
                continue
            
            try:
                request = Request.parse(line)
                response = agent.solve(request)
                print(response.to_json(), flush=True)
                
                if response.task_complete:
                    break
            except Exception as e:
                log_error(f"Error: {e}")
                print(Response.done().to_json(), flush=True)
                break
        
        agent.cleanup()
        
    except KeyboardInterrupt:
        pass
    except Exception as e:
        log_error(f"Fatal: {e}")
        traceback.print_exc(file=sys.stderr)

File: sdk/rust/src/runner.rs
//! Agent runner for Term Challenge.

use std::io::{self, Read, Write};
use crate::{Agent, Request, Response};

/// Log to stderr.
fn log(msg: &str) {
    eprintln!("[agent] {}", msg);
}

/// Run an agent in the Term Challenge harness.
///
/// Reads request from stdin, calls agent.solve(), writes response to stdout.
///
/// ```rust,no_run
/// use term_sdk::{Agent, Request, Response, run};
///
/// struct MyAgent;
///
/// impl Agent for MyAgent {
///     fn solve(&mut self, req: &Request) -> Response {
///         Response::cmd("ls")
///     }
/// }
///
/// fn main() {
///     run(&mut MyAgent);
/// }
/// ```
pub fn run(agent: &mut impl Agent) {
    // Setup
    agent.setup();
    
    // Read input
    let mut input = String::new();
    if let Err(e) = io::stdin().read_to_string(&mut input) {
        log(&format!("Failed to read stdin: {}", e));
        println!("{}", Response::done().to_json());
        return;
    }
    
    let input = input.trim();
    if input.is_empty() {
        log("No input received");
        println!("{}", Response::done().to_json());
        return;
    }
    
    // Parse request
    let request = match Request::parse(input) {
        Ok(req) => req,
        Err(e) => {
            log(&format!("Invalid JSON: {}", e));
            println!("{}", Response::done().to_json());
            return;
        }
    };
    
    log(&format!("Step {}: {}...", request.step, &request.instruction.chars().take(50).collect::<String>()));
    
    // Solve
    let response = agent.solve(&request);
    
    // Output
    println!("{}", response.to_json());
    io::stdout().flush().ok();
    
    // Cleanup
    agent.cleanup();
}

/// Run agent in loop mode (for testing).
pub fn run_loop(agent: &mut impl Agent) {
    use std::io::BufRead;
    
    agent.setup();
    
    let stdin = io::stdin();
    for line in stdin.lock().lines() {
        let line = match line {
            Ok(l) => l,
            Err(_) => break,
        };
        
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        
        let request = match Request::parse(line) {
            Ok(req) => req,
            Err(e) => {
                log(&format!("Invalid JSON: {}", e));
                println!("{}", Response::done().to_json());
                break;
            }
        };
        
        let response = agent.solve(&request);
        println!("{}", response.to_json());
        io::stdout().flush().ok();
        
        if response.task_complete {
            break;
        }
    }
    
    agent.cleanup();
}
